dev/com.ibm.ws.microprofile.config.1.4/src/com/ibm/ws/microprofile/config14/impl/ConfigAccessorImpl.java

```diff
@@ -10,213 +10,177 @@
+                    }
+                } else {
+                    sourcedValue = this.cachedValue;
                 }
-                String variableName = part2.substring(Config14Constants.EVAL_START_TOKEN.length(), evalEndIdx);
-                //System.out.println("VAR: " + variableName);
-                String value = config.getValue(variableName, String.class);
-                //System.out.println("VAL: " + value);
-
-                String part3 = part2.substring(evalEndIdx + Config14Constants.EVAL_END_TOKEN.length());
-                String part3resolved = resolve(config, part3);
-
-                //System.out.println("PT1: " + part1);
-                //System.out.println("PT3: " + part3resolved);
-                resolved = part1 + value + part3resolved;
-                //System.out.println("RES: " + resolved);
-                evalStartIdx = resolved.indexOf(Config14Constants.EVAL_START_TOKEN);
             }
+        } else {
+            sourcedValue = snapshot.getSourcedValue(getPropertyName());
         }
-        return resolved;
+        return sourcedValue;
     }
 
-    public String getResolved(boolean optional) {
-        String value = (String) this.config.getValue(this.propertyName, String.class, optional);
-        if (this.evaluateVariables) {
-            value = resolve(this.config, value);
-        }
+    @Override
+    public SourcedValue getSourcedValue() {
+        SourcedValue value = getSourcedValue(null);
         return value;
     }
 
-    /** {@inheritDoc} */
-    @Override
-    public T getValue() {
-        String resolved = getResolved(false);
-        T value = config.convertValue(resolved, this.conversionType);
-        return value;
+    //must be called within synchronized(this) block
+    private boolean checkCache() {
+        if (this.cacheValid) {
+            long now = System.nanoTime();
+            if ((now - this.cacheExpiryTime) > 0) {
+                this.cacheValid = false;
+            }
+        }
+        return this.cacheValid;
     }
 
     /** {@inheritDoc} */
     @Override
-    public T getValue(ConfigSnapshot configSnapshot) {
-        ConfigSnapshotImpl snapshotImpl = (ConfigSnapshotImpl) configSnapshot;
-        String resolved = snapshotImpl.getResolvedValue(getPropertyName(), false);
-        T value = config.convertValue(resolved, this.conversionType);
-        return value;
+    public String getPropertyName() {
+        return propertyNames.get(propertyNames.size() - 1); //last one is always the base
     }
 
     /** {@inheritDoc} */
     @Override
-    public Optional<T> getOptionalValue() {
-        String resolved = getResolved(true);
-        T value = null;
-        if (resolved != null) {
-            value = config.convertValue(resolved, this.conversionType);
+    public String getResolvedPropertyName() {
+        String resolvedPropertyName;
+        synchronized (this) {
+            resolvedPropertyName = mostRecentPropertyName;
```

Comment:
I don't think this will work unless the user has called `getValue()` previously.

Reply:
Correct, the api javadoc currently says `This should be called only after calling {@link #getValue()} otherwise the value is undefined (but likely null).`